devito/operator.py

```diff
@@ -117,24 +119,66 @@ def __init__(self, expressions, **kwargs):
         # Introduce the required symbol declarations
         nodes = iet_insert_C_decls(dle_state.nodes, self.func_table)
 
-        # Initialise ArgumentEngine
-        self.argument_engine = ArgumentEngine(clusters.ispace, parameters,
-                                              self.dle_arguments)
+        # Insert data and pointer casts for array parameters and profiling structs
+        functions = FindSymbols('symbolics').visit(nodes)
+        casts = [ArrayCast(f) for f in functions if f.is_Tensor and f._mem_external]
+        profiler = Object(self.profiler.name, self.profiler.dtype, self.profiler.new)
+        casts.append(PointerCast(profiler))
+        nodes = (List(body=casts), nodes)
 
-        parameters = self.argument_engine.arguments
+        # Derive parameters as symbols not defined in the kernel itself
+        parameters = derive_parameters(nodes)
+
+        # Filter all internally-allocated temporary `Array` types
```

Comment:
Well, not really as all the DLE transformations have to happen for the IET to be in its final shape and the whole point of this is to use the actual final IET and not lug around assumptions from a previous IR state.

Reply:
But parameter derivation is about more than just function collection - it requires knowledge of who defines (all types of) variables. This means, all things that adjust variable definitions, like elementals or `insert_C_decls` have to have happened before we can safely derive the final `parameter`s. 

What this "hack" is really doing is effectively "predicting" what the `Allocator` will do, because `Allocator` only inserts raw CGen format for local temporary allocations, which cannot be read from the IET state. In general, I don't like this approach of "predicting what another part of the code will do", since this is exactly what caused the inaccurate over-estimation of parameters in the first place (and thus got us into this mess). So instead, I think "the right thing to do" here is to adjust the `Allocator` to insert local array allocations into the IET, allowing us to actually use a fixed state for derivation instead of a "previous state + assumptions".